File: python/src/contriwork_notifications/adapters/slack.py
"""Slack incoming-webhook adapter — HTTPS POST with a JSON body."""

from __future__ import annotations

from typing import Any

import httpx

from ..adapter import AdapterDeliverResult, AdapterStatus
from ..errors import ErrorCode
from ..payload import Payload
from ..severity import Severity

_DEFAULT_TIMEOUT_S = 10.0

_GENERIC_FAILURE = "Notification delivery failed"
_GENERIC_AUTH = "Authentication rejected"


class SlackWebhookAdapter:
    """Slack incoming-webhook delivery via HTTPS POST.

    The webhook URL is opaque to the package (consumer's responsibility to
    load from its own config source). The body is sent as plain text using
    Slack's mrkdwn-friendly conventions (``*bold title*`` + body); the
    package does not produce Block Kit payloads in v0.1.0.

    Slack incoming webhooks have no native silent flag. ``silent`` is a
    documented no-op; CONTRACT.md §Quiet hours flags this explicitly.
    """

    def __init__(
        self,
        webhook_url: str,
        *,
        name: str = "slack-webhook",
        timeout_seconds: float = _DEFAULT_TIMEOUT_S,
    ) -> None:
        self._webhook_url = webhook_url
        self.name = name
        self._timeout = timeout_seconds

    async def is_available(self) -> bool:
        return bool(self._webhook_url)

    async def deliver(
        self,
        severity: Severity,
        payload: Payload,
        silent: bool,
    ) -> AdapterDeliverResult:
        del severity, silent  # see class docstring; silent is a no-op
        text = f"*{payload.title}*\n{payload.body}"
        if payload.url is not None:
            label = payload.url_title or payload.url
            text += f"\n<{payload.url}|{label}>"
        body: dict[str, Any] = {"text": text}

        try:
            async with httpx.AsyncClient(timeout=self._timeout) as client:
                response = await client.post(self._webhook_url, json=body)
        except httpx.TimeoutException:
            return AdapterDeliverResult(
                status=AdapterStatus.RETRIABLE_FAILURE,
                error_code=ErrorCode.TIMEOUT,
                detail=_GENERIC_FAILURE,
            )
        except httpx.HTTPError:
            return AdapterDeliverResult(
                status=AdapterStatus.RETRIABLE_FAILURE,
                error_code=ErrorCode.UPSTREAM_ERROR,
                detail=_GENERIC_FAILURE,
            )

        return _map_response(response.status_code)


def _map_response(status_code: int) -> AdapterDeliverResult:
    if 200 <= status_code < 300:
        return AdapterDeliverResult(status=AdapterStatus.DELIVERED)
    if status_code == 429:
        return AdapterDeliverResult(
            status=AdapterStatus.RETRIABLE_FAILURE,
            error_code=ErrorCode.RATE_LIMITED,
            detail=_GENERIC_FAILURE,
        )
    if status_code in (401, 403):
        return AdapterDeliverResult(
            status=AdapterStatus.PERMANENT_FAILURE,
            error_code=ErrorCode.AUTH_FAILED,
            detail=_GENERIC_AUTH,
        )
    if 400 <= status_code < 500:
        return AdapterDeliverResult(
            status=AdapterStatus.PERMANENT_FAILURE,
            error_code=ErrorCode.ADAPTER_INVALID_PAYLOAD,
            detail=_GENERIC_FAILURE,
        )
    return AdapterDeliverResult(
        status=AdapterStatus.RETRIABLE_FAILURE,
        error_code=ErrorCode.UPSTREAM_ERROR,
        detail=_GENERIC_FAILURE,
    )

File: typescript/src/adapters/slack.ts
import type { Adapter, AdapterDeliverResult } from "../adapter";
import { AdapterStatus } from "../adapter";
import { ErrorCode } from "../errors";
import type { Payload } from "../payload";
import type { Severity } from "../severity";

const DEFAULT_TIMEOUT_MS = 10_000;
const GENERIC_FAILURE = "Notification delivery failed";
const GENERIC_AUTH = "Authentication rejected";

export interface SlackWebhookOptions {
  readonly webhookUrl: string;
  readonly name?: string;
  readonly timeoutMs?: number;
  readonly fetchImpl?: typeof fetch;
}

/**
 * Slack incoming-webhook delivery. Plain mrkdwn body (`*title*` + body,
 * optional `<url|label>` link). Slack incoming webhooks have no native
 * silent flag; `silent` is a documented no-op.
 */
export class SlackWebhookAdapter implements Adapter {
  readonly name: string;
  private readonly webhookUrl: string;
  private readonly timeoutMs: number;
  private readonly fetchImpl: typeof fetch;

  constructor(options: SlackWebhookOptions) {
    this.webhookUrl = options.webhookUrl;
    this.name = options.name ?? "slack-webhook";
    this.timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    this.fetchImpl = options.fetchImpl ?? globalThis.fetch.bind(globalThis);
  }

  async isAvailable(): Promise<boolean> {
    return Boolean(this.webhookUrl);
  }

  async deliver(
    _severity: Severity,
    payload: Payload,
    _silent: boolean,
  ): Promise<AdapterDeliverResult> {
    let text = `*${payload.title}*\n${payload.body}`;
    if (payload.url !== undefined) {
      const label = payload.urlTitle ?? payload.url;
      text += `\n<${payload.url}|${label}>`;
    }
    const body = JSON.stringify({ text });

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    let response: Response;
    try {
      response = await this.fetchImpl(this.webhookUrl, {
        method: "POST",
        headers: { "content-type": "application/json" },
        body,
        signal: controller.signal,
      });
    } catch {
      if (controller.signal.aborted) {
        return {
          status: AdapterStatus.RetriableFailure,
          errorCode: ErrorCode.Timeout,
          detail: GENERIC_FAILURE,
        };
      }
      return {
        status: AdapterStatus.RetriableFailure,
        errorCode: ErrorCode.UpstreamError,
        detail: GENERIC_FAILURE,
      };
    } finally {
      clearTimeout(timer);
    }

    return mapResponse(response.status);
  }
}

function mapResponse(statusCode: number): AdapterDeliverResult {
  if (statusCode >= 200 && statusCode < 300) {
    return { status: AdapterStatus.Delivered };
  }
  if (statusCode === 429) {
    return {
      status: AdapterStatus.RetriableFailure,
      errorCode: ErrorCode.RateLimited,
      detail: GENERIC_FAILURE,
    };
  }
  if (statusCode === 401 || statusCode === 403) {
    return {
      status: AdapterStatus.PermanentFailure,
      errorCode: ErrorCode.AuthFailed,
      detail: GENERIC_AUTH,
    };
  }
  if (statusCode >= 400 && statusCode < 500) {
    return {
      status: AdapterStatus.PermanentFailure,
      errorCode: ErrorCode.AdapterInvalidPayload,
      detail: GENERIC_FAILURE,
    };
  }
  return {
    status: AdapterStatus.RetriableFailure,
    errorCode: ErrorCode.UpstreamError,
    detail: GENERIC_FAILURE,
  };
}
